src/net_processing.cpp

```diff
@@ -1723,6 +1729,9 @@ void PeerManagerImpl::BlockConnected(const std::shared_ptr<const CBlock>& pblock
             m_txrequest.ForgetTxHash(ptx->GetWitnessHash());
         }
     }
+    if (m_chainman.ActiveChainstate().IsInitialBlockDownload()) {
```

Comment:
I would think only reduce the timeout if the block was in fact downloaded in less time (than the would-be *new* timeout).

Reply:
I like about the current solution that it is very simple, prevents the node from getting stuck, and doesn't require additional bookkeeping of historical block download times.

I think your suggestion wouldn't be completely straightforward to implement: The block being connected here might have been downloaded some time in the past, saved to disk, but only connected now (as a result of its predecessor being connected). So we'd need some data structure to keep track of download times for not-yet-connected blocks and add/remove entries from it during IBD.

If we did this, it would help us cycle through less peers in situations where we assign multiple blocks to a peer and halving the timeout after successfully downloading a block would lead to a stalling situation again - but note that there are also other sources of disconnections that could be improved if we kept track of this kind data during IBD: E.g. if doubling the timeout is not sufficient and we'd need to 4x or 8x it.

So if we want something better but more complicated (with bookkeeping), my feeling is that we should go for another approach altogether, like basing the stalling timeout on a running average over the last received block times from multiple peers instead of a doubling/halving approach.

Reply:
With this current code, it will stall, get a block, stall, get a block, stall, etc repeatedly...

Reply:
Okay

Reply:
> With this current code, it will stall, get a block, stall, get a block, stall, etc repeatedly...

Yes, but only a few times until the blocks preventing the tip from moving are downloaded, then the tip advances by connecting the large number of stashed blocks from the 1024 window, ending the stalling situation. If every peer is equally slow, it doesn't matter if the download a block in 2s or 1 minutes from the viewpoint of the stalling logic.

Reply:
Maybe instead of halving the timeout, we should, on block connection, multiply it with a factor `0.5 > f > 1` to let it go back slower?

Reply:
I agree it seems better to decrease it slowly. If it was a single slow peer, then there would be many blocks coming on time afterwards.

Reply:
With the latest push, I decrease it by a factor 0.85 with each connected block.